file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/LazyAttribute.kt

```diff
@@ -1,20 +1,63 @@
 package com.pkware.filesystem.attributecaching
 
-import kotlin.reflect.KProperty
-
 /**
- * Stores a mutable attribute that is lazily calculated via [initializer] whenever the value is `null` and accessed.
+ * Stores a mutable attribute [value] that is lazily calculated via [initializer] whenever the value is `null` and
+ * accessed.
+ *
+ * Can check the initialization status of the attribute via [initialized].
+ *
+ * If the mutable attribute [value] is set to `null` its [initialized] status is set to `false`.
+ *
  */
-internal class LazyAttribute<T>(private val initializer: () -> T?) {
+public class LazyAttribute<T>(private val initializer: () -> T?) {
 
-    private var value: T? = null
+    private var lazyValue: T? = null
 
-    operator fun getValue(thisRef: Any?, property: KProperty<*>): T? {
-        if (value == null) value = initializer()
-        return value
-    }
+    /**
+     * Shows whether this [LazyAttribute] is initialized or not. `true` for initialized, default is `false`.
+     */
+    public var initialized: Boolean = false
```

Comment:
IMO this variable should be private, though I recognize the value in being able to assert against it.  We can easily test the expected behavior of `initialized` and `getValue()` via unit tests that take in some kind of initializer. **Most** of the assertions against `initialized` in tests don't serve great value, look like boilerplate, and detract from the intended purpose of the test.

If we don't have confidence that we're doing the right thing in scenarios, I think something like jfrunit can give us equal (if not greater) confidence that we're doing the right thing

Reply:
Im not going to write lazyAttribute unit tests for code that was just migrated from cachingPath just because you and ryan wanted to play code golf with it. Moving the code to lazyAttribute did not result in any performance gains and was purely a code golf change.

I am leaving the code and tests as they are to test that the attributes are initialized in the proper way.

It's either left as is or i will revert my most recent commit on this pr to the previous one which I know works and is performant other_iops wise.

Reply:
the initialize checks could be used outside of testing, its just that testing is the only area where it is used currently

Reply:
_could_ and _should_ are 2 different things. At the very least, this should only be internal.

There is literally no good reason why a user of this library would need to check it outside of testing purposes, which they could easily do within the confines of a debugger.

Possible use cases:

1. **PROBLEM**: They want to see if it's cached before setting it on the attributes. **SOLUTION**: Just expose a setter that does this check. It's a far more helpful method than forcing callers to check
2. **PROBLEM**: They want to see if it's cached **SOLUTION**: Ignore them. We literally write back any time they call `set`. It doesn't _matter_ if it's cached or not. It's literally a useless boolean.

Reply:
this is why i disagreed with the coupling of everything into lazy attribute in the first place, it just code golfs it all into a nicer looking package but doesn't functionally do anything.

I want to keep the initialized property exposed but keep the value private

Reply:
We didn't really come to any resolution on this one. 

> Im not going to write lazyAttribute unit tests for code that was just migrated from cachingPath just because you and ryan wanted to play code golf with it. Moving the code to lazyAttribute did not result in any performance gains and was purely a code golf change.

There's a lot to unpack here.

First, I find this to be a misrepresentation of what we suggested. This was not code golf, this was adhering to DRY (for the uninitiated: Don't Repeat Yourself). We removed 5+ if statements from the flow of another method. We avoided 5 extra variables. We moved boilerplate logic to a single method. We simplified the readability. We made it so it _can_ be unit tested. We tried suggesting a basic improvement in _design_.

Second, this particular suggestion has nothing to do with code golf. It's about exposing only the most essential variables as a library. The point was hardly a subjective one, it was bringing attention to a bit of a design smell.

Third, I think DRY also applies to testing. You're basically testing the responsibility of `LazyAttribute` in your tests because there wasn't a unit to verify that behavior at the time you wrote them. It was just a suggestion!

Fourth, and most importantly, Protesting so strongly against a point when someone literally says "To be clear, this isn't a hill I'm looking to die on, it's a pretty minor detail" really doesn't give an impression that you're even trying to listen to feedback. Your repeated resistance to the suggestions of basic good development practices really mutes the strength of the good arguments that you do have.  It gives a tone of hard lines being drawn in the sand that you're unwilling to cross no matter how rational the point and definitely saps the patience you would otherwise be owed from others. I apologize for matching the tone which I perceived to read on this PR.

Reply:
To begin I'm pushing back so hard on these changes because very little changed from 1.0.2 to 1.0.3 and this changes base functionality that I have had in the file attribute cache since we swapped to lazy fields over an actual cache.

I came back with the tone in my replies because PR feedback went like this: this was a relatively simple PR with some minor changes (1.0.2 was approved where major changes took place), current pr feedback to implement change A (consolidate variables to another class), then all of a sudden every major implementation detail around how attributes are saved/cached/done is questioned and analyzed. Outside of the lazyAttribute refactor, every other suggested change makes it seem like nothing was really reviewed in the 1.0.2 PR or even the 1.0.1 PR and suddenly now it's a concern for the sake of being a concern. 

I understand iterative design and changes are welcome, however the manner these changes have been presented in when considering past pr feedback makes me feel like previous prs were not reviewed properly or not enough questions were asked then. Thus, I push back with asking why and justifications for not making such changes.

I wanted cache status checking/initialization to remain the same as it has been but be used for each attribute, then I realized there was an issue with copyAttributesTo and it's current incarnation where checking the attributes of `other` caused it to be cached and thus make an IOPS hit. The copyAttributesTo change was the only functional one made beyond removing all of the original init logic for this PR. 

> First, I find this to be a misrepresentation of what we suggested. This was not code golf, this was adhering to DRY (for the uninitiated: Don't Repeat Yourself). We removed 5+ if statements from the flow of another method. We avoided 5 extra variables. We moved boilerplate logic to a single method. We simplified the readability. We made it so it can be unit tested. We tried suggesting a basic improvement in design.

It still is 99% code golf, the variables were named simply to correspond to the cache status of each attribute, and the solution was very reasonable and readable- you could check the status of every attribute with its corresponding variable. The DRY implementation of LazyAttribute also made testing of the separate cache fields alot more unreadable to me (with the long chaining of variable names). 

There is also extra work involved with unit testing combined code in lazyAttribute that before was properly tested when things were not consolidated. This all being said I implemented the changes as laid out, and asked one question about how will I keep variables properly exposed/not exposed- that was never addressed or answered `question with this is how to i keep the lazy attribute itself private and the cached boolean property public with a private set?`. Throughout the entire design I wanted to keep the cache\init status exposed for each attribute. I disagreed/still disagree with keeping cache\init status private to the lazy attribute class as it helps alot with problems attribute caching that are unknown until they are known.

> Second, this particular suggestion has nothing to do with code golf. It's about exposing only the most essential variables as a library. The point was hardly a subjective one, it was bringing attention to a bit of a design smell.

The lazy attribute refactor may have been a design smell and that's been implemented. However, I think that the cache\init status of each attribute is an essential feature of this library, just like it was good to have with the non on-demand cache. Again, I feel like if this was such an issue it should have been brought up previous prs (namely feedback around cache\initialized status).

> Third, I think DRY also applies to testing. You're basically testing the responsibility of LazyAttribute in your tests because there wasn't a unit to verify that behavior at the time you wrote them. It was just a suggestion!

Before the class/variable consolidation there wasn't a need to unit test Lazy Attribute but the refactor caused the need for it and I figured I could still handle that testing within the Filesystem tests as those are mainly end-to-end integration tests (they test the filesystem + attribute caching path + the provider). I agreed with most of the initial suggestions made on this PR, including the lazy attribute refactor, after some justification. Current outstanding suggestions do not seem like suggestions and more like "implement this, this way, or never get pr approval".

> Fourth, and most importantly, Protesting so strongly against a point when someone literally says "To be clear, this isn't a hill I'm looking to die on, it's a pretty minor detail" really doesn't give an impression that you're even trying to listen to feedback. Your repeated resistance to the suggestions of basic good development practices really mutes the strength of the good arguments that you do have. It gives a tone of hard lines being drawn in the sand that you're unwilling to cross no matter how rational the point and definitely saps the patience you would otherwise be owed from others. I apologize for matching the tone which I perceived to read on this PR.

I was frustrated again by the fact that it felt like this change could have been brought up in the 1.0.2 PRs or a previous PR to this one. I only offered such resistance when it felt like my original design was being suddenly co-op'd (after having multiple releases and PRs where it was considered fine) in a way that I dont agree with as well. I figured my response would also make this a hill to not die on and just compromise without a forced design change on my end in this situation, but it's been very much escalated from that based on your responses.
My tone in responses to this PR was fine until extra suggestions were made to change more long standing design decisions that should have been addressed in previous PRs/releases. 
It's been a consistent theme that I offer changes, everyone signs off on them, then suddenly in PR I have to change the entire design; or even when design changes get approved in previous releases/prs they need to be suddenly changed now.